src/Featurizers/ShortGrainDropperFeaturizer.h

```diff
@@ -0,0 +1,255 @@
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using BaseType                          = TransformerEstimator<std::vector<std::string>, bool>;
+    using TransformerType                   = ShortGrainDropperTransformer;
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    ShortGrainDropperEstimator(
+        AnnotationMapsPtr pAllColumnAnnotations,
+        size_t colIndex,
+        std::uint32_t windowSize,
+        std::vector<std::uint32_t> lags,
+        std::uint32_t maxHorizon,
+        nonstd::optional<std::uint32_t> cv
+    );
+    ~ShortGrainDropperEstimator(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(ShortGrainDropperEstimator);
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using GrainsSet                         = ShortGrainDropperTransformer::GrainsSet;
+    using GrainsMap                         = std::unordered_map<std::string, std::uint32_t>;
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    size_t const                            _colIndex;
+    std::uint32_t const                     _minPoints;
+    GrainsSet                               _grainsToDrop;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+    bool begin_training_impl(void) override;
+
+    // MSVC has problems when the declaration and definition are separated
+    FitResult fit_impl(typename BaseType::InputType const *pBuffer, size_t cElements) override {
+
+        InputType const * const                 pEndBuffer(pBuffer + cElements);
+
+        GrainsMap groupByGrains;
+        while(pBuffer != pEndBuffer) {
+            std::string pBufferStr = Traits<decltype(*pBuffer)>::ToString(*pBuffer);
+            GrainsMap::iterator grainsMapIter(groupByGrains.find(pBufferStr));
+            if (grainsMapIter != groupByGrains.end())
+                ++grainsMapIter->second;
+            else
+                groupByGrains.emplace(pBufferStr, 1);
+            ++pBuffer;
+        }
+
+        for (GrainsMap::value_type const & groupByGrainsElement : groupByGrains) {
+            if (groupByGrainsElement.second <= _minPoints)
+                _grainsToDrop.emplace(std::move(groupByGrainsElement.first));
+        }
```

Comment:
This logic doesn't feel right to me, as it seems like we need to maintain the count across all training and only calculate the grains to drop once training is complete.

Reply:
The logic here is exactly calculating the grains to drop when training is complete, as the signature of fit_impl() is fit_impl(typename BaseType::InputType const *pBuffer, size_t cElements)

Reply:
Let's discuss this in person to make sure that we are on the same page.

---
In reply to: [384062585](https://github.com/microsoft/FeaturizersLibrary/pull/150#discussion_r384062585) [](ancestors = 384062585)